app/models/host/base.rb

```diff
@@ -45,12 +45,18 @@ class Base < ApplicationRecord
     default_scope -> { where(taxonomy_conditions) }
 
     def self.taxonomy_conditions
-      org = Organization.expand(Organization.current)
-      loc = Location.expand(Location.current)
       conditions = {}
-      conditions[:organization_id] = Array(org).map { |o| o.subtree_ids }.flatten.uniq unless org.nil?
-      conditions[:location_id] = Array(loc).map { |l| l.subtree_ids }.flatten.uniq unless loc.nil?
-      conditions
+      if Organization.current.nil? && User.current.present? && !User.current.admin?
+        conditions[:organization_id] = User.current.organization_and_child_ids
+      else
+        conditions[:organization_id] = Organization.current&.subtree_ids
+      end
+      if Location.current.nil? && User.current.present? && !User.current.admin?
+        conditions[:location_id] = User.current.location_and_child_ids
+      else
+        conditions[:location_id] = Location.current&.subtree_ids
+      end
+      conditions.compact
```

Comment:
Why compacting is needed? It could be moved into the condition and `location_and_child_ids` method so it's also cached.

Reply:
the compact is needed when one of the taxonomies is `nil` and there is no current user or the current user is admin, in which case we don't want the condition for that taxonomy.

Reply:
Just got caught by this, code above can result in conditions[:organization] being nil, which AR would translate to hosts.organization_id IS NULL which would be incorrect. With this it does not happen. A comment would be nice, but I think it's alright for now.